README.md
# Entropy

- [Entropy](#entropy)
	- [Description](#description)
	- [Usage](#usage)
	- [Why](#why)
	- [References](#references)

## Description

This package will calculate the entropy bits from a given string. This is a metric, unintended to be used by itself, to make basic credentials more secure. It is a good idea to use this in conjunction with an updated password hashing algorithm such as [Argon2](https://github.com/P-H-C/phc-winner-argon2). The Go version can be found here: [Argon2 Go Package](golang.org/x/crypto/argon2).

## Usage

```go
package main

import (
	"fmt"

	"github.com/adroge/entropy"
)

func main() {
	password := "monkey"
	result, err := entropy.Calculate(password)
	if err != nil {
		fmt.Println("error: ", err)
		return
	}
	fmt.Printf("Strength of \"%s\" is %s, with entropy: %f\n", password, result, result.Bits)
}
```

```sh
$ go run main.go
Strength of "monkey" is very weak, with entropy: 28.202638
```

It's possible to change the alphabets and entropy strength bounds by utilizing the Alphabet(...) and Bounds(...) functions. If you wanted to include a subset of the Latin-1 Supplement characters, you could do it with Alphabet(...).

Unit tests are a good source of usage.

## Why

This module is intended to be used for backend development where a service needs to quickly validate the entropy of a password.

There are other packages out there that do similar things, but they were either, in my opinion, overkill, or didn't fit the need I had.

## References

These aren't the only places, or even the original places, where the formula used can be found, but for my own reference I used the following:

- [How to Calculate Password Entropy?](https://generatepasswords.org/how-to-calculate-entropy/)
- [Password Security: Complexity vs. Length](http://resources.infosecinstitute.com/password-security-complexity-vs-length/)
- [Entropy as a measure of password strength](https://en.wikipedia.org/wiki/Password_strength#Entropy_as_a_measure_of_password_strength)


CHANGELOG.md
# Changelog

- [Changelog](#changelog)
  - [About](#about)
  - [[v1.0.1]](#v101)
    - [Changed v1.0.1](#changed-v101)
  - [[v1.0.0]](#v100)

## About

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.0.1]

### Changed v1.0.1

- Go 1.18.
- entropy strength to a type internally.
- SetEntropyMock -> SetMock - matches documentation.
- Documentation on usage updated.

## [v1.0.0]

Initial release.


entropy_test.go
package entropy_test

import (
	"errors"
	"testing"

	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"

	"github.com/adroge/entropy"
	"github.com/adroge/entropy/mock_entropy"
)

type EntropyTestSuite struct {
	suite.Suite
}

type EntropyMockTestSuite struct {
	suite.Suite
}

func TestExampleTestSuite(t *testing.T) {
	suite.Run(t, new(EntropyTestSuite))
	suite.Run(t, new(EntropyMockTestSuite))
}

func (test *EntropyTestSuite) SetupTest() {
	alphabets := []string{
		`abcdefghijklmnopqrstuvwxyz`,
		`ABCDEFGHIJKLMNOPQRSTUVWXYZ`,
		`1234567890`,
		`!@#$%^&*`,
		`()[]{}<>`,
		`~-_=+|;:',./?\ "` + "`",
	}
	err := entropy.Alphabets(alphabets)
	require.Nil(test.T(), err)
	err = entropy.Bounds(30.0, 40.0, 60.0, 127.0)
	require.Nil(test.T(), err)
	err = entropy.Descriptions(entropy.DescriptionTags{
		Invalid:    "invalid",
		VeryWeak:   "very weak",
		Weak:       "weak",
		Reasonable: "reasonable",
		Strong:     "strong",
		VeryStrong: "very strong"})
	require.Nil(test.T(), err)
}

func (test *EntropyTestSuite) TestCalculateEmptyString() {
	result, err := entropy.Calculate("")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), "very weak", result.String())
	assert.Equal(test.T(), entropy.VeryWeak, result.Evaluation)
}

func (test *EntropyTestSuite) TestVeryWeak() {
	result, err := entropy.Calculate("abcd")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 18.801758872564367, result.Bits)
	assert.Equal(test.T(), "very weak", result.String())
	assert.Equal(test.T(), entropy.VeryWeak, result.Evaluation)
}

func (test *EntropyTestSuite) TestWeak() {
	result, err := entropy.Calculate("abcdefgh")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 37.603517745128734, result.Bits)
	assert.Equal(test.T(), "weak", result.String())
	assert.Equal(test.T(), entropy.Weak, result.Evaluation)
}

func (test *EntropyTestSuite) TestReasonable() {
	result, err := entropy.Calculate("abcdEFGH")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 45.603517745128734, result.Bits)
	assert.Equal(test.T(), "reasonable", result.String())
	assert.Equal(test.T(), entropy.Reasonable, result.Evaluation)
}

func (test *EntropyTestSuite) TestReasonableWithNumber() {
	result, err := entropy.Calculate("abcdE1GH")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 47.633570483095, result.Bits)
	assert.Equal(test.T(), "reasonable", result.String())
	assert.Equal(test.T(), entropy.Reasonable, result.Evaluation)
}

func (test *EntropyTestSuite) TestReasonableWithNumberAndSymbol() {
	result, err := entropy.Calculate("ab$dE1GH")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 49.03426413555973, result.Bits)
	assert.Equal(test.T(), "reasonable", result.String())
	assert.Equal(test.T(), entropy.Reasonable, result.Evaluation)
}

func (test *EntropyTestSuite) TestStrong() {
	result, err := entropy.Calculate("Bc8$5yjvK>8Y")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 75.42482662634698, result.Bits)
	assert.Equal(test.T(), "strong", result.String())
	assert.Equal(test.T(), entropy.Strong, result.Evaluation)
}

func (test *EntropyTestSuite) TestVeryStrong() {
	result, err := entropy.Calculate("J90;1]6rtpZ4ny;.EZ:wW")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 135.3018134128233, result.Bits)
	assert.Equal(test.T(), "very strong", result.String())
	assert.Equal(test.T(), entropy.VeryStrong, result.Evaluation)
}

func (test *EntropyTestSuite) TestRuneNotInAlphabet() {
	_, err := entropy.Calculate("zufälliges Passwort")
	require.NotNil(test.T(), err)
	assert.True(test.T(), errors.Is(err, entropy.ErrUnexpectedRune))
	assert.Equal(test.T(), "unexpected rune in input: ä", err.Error())
}

func (test *EntropyTestSuite) TestRuneInLatinAlphabet() {
	alphabets := []string{
		`äabcdefghijklmnopqrstuvwxyz`,
		`ABCDEFGHIJKLMNOPQRSTUVWXYZ`,
		`~-_=+|;:',./? \"` + "`",
	}
	err := entropy.Alphabets(alphabets)
	assert.Nil(test.T(), err)
	result, err := entropy.Calculate("zufälliges Passwort")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), "strong", result.String())
	assert.Equal(test.T(), entropy.Strong, result.Evaluation)
}

func (test *EntropyTestSuite) TestAlphabetsEmpty() {
	err := entropy.Alphabets([]string{})
	require.NotNil(test.T(), err)
	assert.True(test.T(), errors.Is(err, entropy.ErrInvalidAlphabet))
}

func (test *EntropyTestSuite) TestChangeBounds() {
	err := entropy.Bounds(5, 10, 15, 20)
	assert.Nil(test.T(), err)
	result, err := entropy.Calculate("monkey")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), 28.202638308846552, result.Bits)
	assert.Equal(test.T(), "very strong", result.String())
	assert.Equal(test.T(), entropy.VeryStrong, result.Evaluation)
}

func (test *EntropyTestSuite) TestChangeBoundsBadInput() {
	err := entropy.Bounds(50, 10, 15, 20)
	assert.NotNil(test.T(), err)
	assert.True(test.T(), errors.Is(err, entropy.ErrInvalidEntropy))
}

func (test *EntropyTestSuite) TestChangeDescriptionInvalid() {
	err := entropy.Descriptions(entropy.DescriptionTags{"", "", "", "", "", ""})
	assert.NotNil(test.T(), err)
	assert.True(test.T(), errors.Is(err, entropy.ErrInvalidDescription))
}

func (test *EntropyTestSuite) TestChangeDescriptionVeryWeak() {
	err := entropy.Descriptions(entropy.DescriptionTags{"invalid", "pitiful", "laughable", "passable", "impressive", "ludicrous"})
	assert.Nil(test.T(), err)
	result, err := entropy.Calculate("monkey")
	assert.Nil(test.T(), err)
	assert.Equal(test.T(), "pitiful", result.String())
	assert.Equal(test.T(), entropy.VeryWeak, result.Evaluation)
}

func (test *EntropyTestSuite) TestEntropyBounds() {
	bounds := entropy.EntropyBounds()
	assert.EqualValues(test.T(), []float64{30.0, 40.0, 60.0, 127.0}, bounds)
}

func (test *EntropyMockTestSuite) TestMock() {
	controller := gomock.NewController(test.T())
	entropyMock := mock_entropy.NewMockFunction(controller)
	entropy.SetMock(entropyMock)
	entropyMock.EXPECT().Calculate(gomock.Any()).Return(entropy.EntropyResult{}, entropy.ErrUnexpectedRune)
	_, err := entropy.Calculate("password")
	require.NotNil(test.T(), err)
	assert.True(test.T(), errors.Is(err, entropy.ErrUnexpectedRune))
}


mock_entropy/mock_entropy.go
// Code generated by MockGen. DO NOT EDIT.
// Source: entropy.go

// Package mock_entropy is a generated GoMock package.
package mock_entropy

import (
	reflect "reflect"

	entropy "github.com/adroge/entropy"
	gomock "github.com/golang/mock/gomock"
)

// MockFunction is a mock of Function interface.
type MockFunction struct {
	ctrl     *gomock.Controller
	recorder *MockFunctionMockRecorder
}

// MockFunctionMockRecorder is the mock recorder for MockFunction.
type MockFunctionMockRecorder struct {
	mock *MockFunction
}

// NewMockFunction creates a new mock instance.
func NewMockFunction(ctrl *gomock.Controller) *MockFunction {
	mock := &MockFunction{ctrl: ctrl}
	mock.recorder = &MockFunctionMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockFunction) EXPECT() *MockFunctionMockRecorder {
	return m.recorder
}

// Alphabets mocks base method.
func (m *MockFunction) Alphabets(newAlphabets []string) error {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "Alphabets", newAlphabets)
	ret0, _ := ret[0].(error)
	return ret0
}

// Alphabets indicates an expected call of Alphabets.
func (mr *MockFunctionMockRecorder) Alphabets(newAlphabets interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Alphabets", reflect.TypeOf((*MockFunction)(nil).Alphabets), newAlphabets)
}

// Bounds mocks base method.
func (m *MockFunction) Bounds(veryWeak, weak, reasonable, strong float64) error {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "Bounds", veryWeak, weak, reasonable, strong)
	ret0, _ := ret[0].(error)
	return ret0
}

// Bounds indicates an expected call of Bounds.
func (mr *MockFunctionMockRecorder) Bounds(veryWeak, weak, reasonable, strong interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Bounds", reflect.TypeOf((*MockFunction)(nil).Bounds), veryWeak, weak, reasonable, strong)
}

// Calculate mocks base method.
func (m *MockFunction) Calculate(input string) (entropy.EntropyResult, error) {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "Calculate", input)
	ret0, _ := ret[0].(entropy.EntropyResult)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}

// Calculate indicates an expected call of Calculate.
func (mr *MockFunctionMockRecorder) Calculate(input interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Calculate", reflect.TypeOf((*MockFunction)(nil).Calculate), input)
}

// Descriptions mocks base method.
func (m *MockFunction) Descriptions(tags entropy.DescriptionTags) error {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "Descriptions", tags)
	ret0, _ := ret[0].(error)
	return ret0
}

// Descriptions indicates an expected call of Descriptions.
func (mr *MockFunctionMockRecorder) Descriptions(tags interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Descriptions", reflect.TypeOf((*MockFunction)(nil).Descriptions), tags)
}

// EntropyBounds mocks base method.
func (m *MockFunction) EntropyBounds() []float64 {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "EntropyBounds")
	ret0, _ := ret[0].([]float64)
	return ret0
}

// EntropyBounds indicates an expected call of EntropyBounds.
func (mr *MockFunctionMockRecorder) EntropyBounds() *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "EntropyBounds", reflect.TypeOf((*MockFunction)(nil).EntropyBounds))
}


entropy.go
// Package entropy is used to calculate the entropy of a string.
//
// This can be used as one of the means to help determine how safe a password is in backend code.
//
// See entropy_test.go for usage.
package entropy

//go:generate mockgen -source entropy.go -destination mock_entropy/mock_entropy.go

import (
	"errors"
	"fmt"
	"math"
)

type entropyStrength int

const (
	VeryWeak entropyStrength = iota
	Weak
	Reasonable
	Strong
	VeryStrong
	Invalid
)

var (
	ErrInvalidAlphabet    = errors.New("invalid alphabet")
	ErrInvalidEntropy     = errors.New("invalid entropy")
	ErrInvalidDescription = errors.New("invalid description")
	ErrUnexpectedRune     = errors.New("unexpected rune in input")

	_Invalid    = "invalid"
	_Weak       = "weak"
	_VeryWeak   = "very weak"
	_Reasonable = "reasonable"
	_Strong     = "strong"
	_VeryStrong = "very strong"

	evaluationWords = map[entropyStrength]*string{
		Invalid:    &_Invalid,
		VeryWeak:   &_VeryWeak,
		Weak:       &_Weak,
		Reasonable: &_Reasonable,
		Strong:     &_Strong,
		VeryStrong: &_VeryStrong,
	}

	runeTypes map[rune]int

	alphabets = []string{
		`abcdefghijklmnopqrstuvwxyz`,
		`ABCDEFGHIJKLMNOPQRSTUVWXYZ`,
		`1234567890`,
		`!@#$%^&*`,
		`()[]{}<>`,
		`~-_=+|;:',./? \"` + "`",
	}

	entropyUpperBounds = [...]float64{
		30.0,
		40.0,
		60.0,
		127.0,
	}
)

// EntropyResult contains the bits, and the evaluation that maps to exported constants
type EntropyResult struct {
	Bits       float64
	Evaluation entropyStrength
}

// DescriptionTags can be used to change the string representation of an entropy value
type DescriptionTags struct {
	Invalid    string
	VeryWeak   string
	Weak       string
	Reasonable string
	Strong     string
	VeryStrong string
}

// String implements the stringer interface
func (er EntropyResult) String() string {
	return *evaluationWords[er.Evaluation]
}

// Function defines an interface that is used for testing
// with the generated mock_entropy code.
type Function interface {
	Alphabets(newAlphabets []string) (err error)
	Bounds(veryWeak, weak, reasonable, strong float64) (err error)
	Calculate(input string) (result EntropyResult, err error)
	Descriptions(tags DescriptionTags) (err error)
	EntropyBounds() (bounds []float64)
}

type functions struct{}

var call Function = functions{}

func init() {
	loadAlphabets()
}

func loadAlphabets() {
	runeTypes = make(map[rune]int)
	for alphabetIndex, alphabet := range alphabets {
		for _, runeCharacter := range alphabet {
			if _, found := runeTypes[runeCharacter]; !found {
				runeTypes[runeCharacter] = alphabetIndex
			}
		}
	}
}

// Alphabets replaces the currently defined alphabets.
func Alphabets(newAlphabets []string) (err error) {
	return call.Alphabets(newAlphabets)
}
func (functions) Alphabets(newAlphabets []string) (err error) {
	if len(newAlphabets) == 0 {
		return ErrInvalidAlphabet
	}
	alphabets = newAlphabets
	loadAlphabets()
	return
}

// Bounds changes the upper ranges for evaluating the strength
// of the calculated entropy.
func Bounds(veryWeak, weak, reasonable, strong float64) (err error) {
	return call.Bounds(veryWeak, weak, reasonable, strong)
}
func (functions) Bounds(veryWeak, weak, reasonable, strong float64) (err error) {
	if veryWeak >= weak || weak >= reasonable || reasonable >= strong {
		return ErrInvalidEntropy
	}
	entropyUpperBounds[VeryWeak] = veryWeak
	entropyUpperBounds[Weak] = weak
	entropyUpperBounds[Reasonable] = reasonable
	entropyUpperBounds[Strong] = strong
	return
}

// Calculate analyzes the entropy of a string.
func Calculate(input string) (result EntropyResult, err error) {
	return call.Calculate(input)
}
func (functions) Calculate(input string) (result EntropyResult, err error) {
	if result.Bits, err = calculateEntropy(input); err != nil {
		return
	}
	result.Evaluation, err = evaluateEntropy(result.Bits)
	return
}

func calculateEntropy(input string) (entropy float64, err error) {
	if len(input) == 0 {
		return
	}
	characterPool := make(map[int]int)
	for _, character := range input {
		if _, found := runeTypes[character]; !found {
			err = fmt.Errorf("%w: %c", ErrUnexpectedRune, character)
			return
		}
		characterPool[runeTypes[character]] = len(alphabets[runeTypes[character]])
	}
	poolSize := 0
	for _, length := range characterPool {
		poolSize += length
	}
	entropy = math.Log2(math.Pow(float64(poolSize), float64(len(input))))
	return
}

func evaluateEntropy(entropy float64) (evaluation entropyStrength, err error) {
	switch {
	case math.IsNaN(entropy) || entropy < 0.0:
		evaluation = Invalid
		err = fmt.Errorf("%w: %f", ErrInvalidEntropy, entropy)
	case entropy <= entropyUpperBounds[VeryWeak]:
		evaluation = VeryWeak
	case entropy <= entropyUpperBounds[Weak]:
		evaluation = Weak
	case entropy <= entropyUpperBounds[Reasonable]:
		evaluation = Reasonable
	case entropy <= entropyUpperBounds[Strong]:
		evaluation = Strong
	default:
		evaluation = VeryStrong
	}
	return
}

// Descriptions changes the string descriptions of the String() evaluation
// of the entropy value.
func Descriptions(tags DescriptionTags) (err error) {
	return call.Descriptions(tags)
}
func (functions) Descriptions(tags DescriptionTags) (err error) {
	if len(tags.Invalid) == 0 || len(tags.VeryWeak) == 0 ||
		len(tags.Weak) == 0 || len(tags.Reasonable) == 0 ||
		len(tags.Strong) == 0 || len(tags.VeryStrong) == 0 {
		return ErrInvalidDescription
	}
	_Invalid = tags.Invalid
	_VeryWeak = tags.VeryWeak
	_Weak = tags.Weak
	_Reasonable = tags.Reasonable
	_Strong = tags.Strong
	_VeryStrong = tags.VeryStrong
	return
}

// EntropyBounds returns an array of bound values used internally.
func EntropyBounds() (bounds []float64) {
	return call.EntropyBounds()
}
func (functions) EntropyBounds() (bounds []float64) {
	bounds = make([]float64, 4)
	copy(bounds, entropyUpperBounds[:])
	return
}
